File: comm/PersistentUnionFind.cpp
/*
 * Persistent Union Find tree.
 * Reference: https://misteer.hatenablog.com/entry/persistentUF
 * Header requirement: algorithm, utility, vector
 * Verified by https://beta.atcoder.jp/contests/agc002/submissions/3185458
 */
class PersistentUnionFind {
public:
  static const int inf = 1e8;
  typedef std::pair<int, int> pii;
  std::vector<int> par, time, rank;
  std::vector<std::vector<pii> > num; // [(time, size of this component)]
  int now;
  PersistentUnionFind(int n):
    par(n), time(n, inf), rank(n, 0), num(n), now(0) {
    for (int i = 0; i < n; ++i) {
      par[i] = i;
      num[i].push_back(pii(-1, 1));
    }
  }
  int root(int x, int t) const {
    while (1) {
      if (time[x] > t) return x;
      x = par[x];
    }
  }
  // returns the current time
  int unite(int x, int y) {
    x = this->root(x, now);
    y = this->root(y, now);
    if (x == y) return now;
    ++now;
    if (rank[x] <= rank[y]) std::swap(x, y);
    par[y] = x;
    time[y] = now;
    rank[x] = std::max(rank[x], rank[y] + 1);
    num[x].push_back(pii(now, num[x].back().second + num[y].back().second));
    return now;
  }
  int size(int x, int t) const {
    x = root(x, t);
    return (upper_bound(num[x].begin(), num[x].end(), pii(t, inf)) - 1)->second;
  }
};
const int PersistentUnionFind::inf;

File: comm/PersistentUnionFind.rs
/*
 * Persistent Union Find tree.
 * Reference: https://misteer.hatenablog.com/entry/persistentUF
 */
struct PersistentUnionFind {
    par: Vec<usize>,
    time: Vec<i32>,
    rank: Vec<i32>,
    num: Vec<Vec<(i32, usize)>>, // [(time, size of this component)]
    now: i32,
}
impl PersistentUnionFind {
    fn new(n: usize) -> Self {
        let mut par = vec![0; n];
        let time = vec![i32::max_value(); n];
        let rank = vec![0; n];
        let num = vec![vec![(-1, 1)]; n];
        for i in 0 .. n {
            par[i] = i;
        }
        PersistentUnionFind {
            par: par,
            time: time,
            rank: rank,
            num: num,
            now: 0
        }
    }
    fn root(&self, mut x: usize, t: i32) -> usize {
        loop {
            if self.time[x] > t { return x; }
            x = self.par[x];
        }
    }
    // returns the current time
    #[allow(dead_code)]
    fn unite(&mut self, x: usize, y: usize) -> i32 {
        let mut now = self.now;
        let mut x = self.root(x, now);
        let mut y = self.root(y, now);
        if x == y { return now; }
        now += 1;
        if self.rank[x] <= self.rank[y] { std::mem::swap(&mut x, &mut y); }
        self.par[y] = x;
        self.time[y] = now;
        self.rank[x] = std::cmp::max(self.rank[x], self.rank[y] + 1);
        let size0 = self.num[x].last().unwrap().1;
        let size1 = self.num[y].last().unwrap().1;
        self.num[x].push((now, size0 + size1));
        self.now = now;
        now
    }
    #[allow(dead_code)]
    fn size(&self, x: usize, t: i32) -> usize {
        let x = self.root(x, t);
        let idx = self.num[x].binary_search(&(t, usize::max_value())).err().unwrap() - 1;
        self.num[x][idx].1
    }
}
